impl/storage.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2020 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package impl
+
+import (
+	"context"
+	"fmt"
+
+	"cloud.google.com/go/spanner"
+	"github.com/google/keytransparency/core/sequencer"
+	"github.com/google/keytransparency/core/water"
+	"github.com/google/keytransparency/impl/mysql/mutationstorage"
+	"github.com/google/keytransparency/impl/spanner/batch"
+	"github.com/google/keytransparency/impl/spanner/directory"
+	"github.com/google/keytransparency/impl/spanner/mutations"
+
+	pb "github.com/google/keytransparency/core/api/v1/keytransparency_go_proto"
+	dinterface "github.com/google/keytransparency/core/directory"
+	ktsql "github.com/google/keytransparency/impl/mysql"
+	sqld "github.com/google/keytransparency/impl/mysql/directory"
+)
+
+// Storage holds an abstract storage implementation
+type Storage struct {
+	Directories dinterface.Storage
+	Logs        interface {
+		sequencer.LogsReader
+
+		// Copied methods from adminserver.LogsAdmin because of duplicate method.
+
+		// AddLogs creates and adds new logs for writing to a directory.
+		AddLogs(ctx context.Context, directoryID string, logIDs ...int64) error
+		// SetWritable enables or disables new writes from going to logID.
+		SetWritable(ctx context.Context, directoryID string, logID int64, enabled bool) error
+
+		// Copied methods from keyserver.MutationLogs
+		SendBatch(ctx context.Context, directoryID string, logID int64, batch []*pb.EntryUpdate) (water.Mark, error)
+	}
```

Comment:
Can you get a Go Native to review this? It looks a bit clunky to me.

Reply:
This is a bit of a known issue: https://github.com/golang/go/issues/6977
https://golang.org/ref/spec#Uniqueness_of_identifiers

I think these could be merged if the interfaces were in the same package, because they are in different packages, they are different identifiers.